Project 9 - Submit #1

MarkALevy opened the pull request.

MarkALevy committed:
load user information from the server

MarkALevy committed:
finish stages 2-6

MarkALevy committed:
feat: add and remove likes

MarkALevy committed:
feat:update profile picture

MarkALevy committed:
Submit #1

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,90 @@
+    return fetch(`${this._baseUrl}/cards/${cardId}`, {
+      method: "DELETE",
+      headers: this._headers,
+    }).then((res) => this._handleRes(res));
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._handleRes)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

MarkALevy replied:
I tried what you suggested, but I get an error while calling the api methods in index.js:
"index.js:44 Uncaught TypeError: Cannot read properties of undefined (reading 'then')"
is there something I need to change in the syntax? 
![image](https://github.com/user-attachments/assets/e9c7c753-dd6d-49ee-8b36-6015caae530b)
![image](https://github.com/user-attachments/assets/b4d88cdb-c348-4508-af1c-b762d7977585)

gennady-bars replied:
> I tried what you suggested, but I get an error while calling the api methods in index.js: "index.js:44 Uncaught TypeError: Cannot read properties of undefined (reading 'then')" is there something I need to change in the syntax?  


You've forgotten `return` before `this._request `

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -28,7 +29,8 @@ export default class PopupWithForm extends Popup {
     this._popupForm.addEventListener("submit", (evt) => {
       evt.preventDefault();
-      this._handleFormSubmit(this._getInputValues());
+      this._submitButton.textContent = "Saving...";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

MarkALevy replied:
where do I get the value of `isLoading`?
do I just define it to be `true` while calling the method in the submit handler?
if so - when does return to `false`?

gennady-bars replied:
Before the request it's `true`, and in `finally` it is `false`  (in `index.js`)

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -8,6 +8,7 @@ export default class PopupWithForm extends Popup {
+    this._submitButton = this._popupForm.querySelector(".modal__save-button");
     this._handleFormSubmit = handleFormSubmit;
   }
 
```
**Could be improved**
You can make a special method for getting a form element from `PopupWithForm` not to duplicate the searching in `index.js`



```js

  getForm() {
    return this._popupForm
  }


```

MarkALevy replied:
I understand how this could generally be more efficient, but I'm not sure I understand what duplicate searching you're referring to in index.js

gennady-bars replied:
> I understand how this could generally be more efficient, but I'm not sure I understand what duplicate searching you're referring to in index.js

Actually, you don't duplicate searching because you get the froms from `document.forms` (it's a nice way). I didn't see that before I left this comment. 

So, everything is fine

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+      userJob: res.about,
+    });
+
+    profileImageEl.setAttribute("src", res.avatar);
```
`UserInfo` should set the entire user data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+    })
+    .finally(() => {
+      profileEditPopup.close();
+      button.textContent = "Save";
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

MarkALevy replied:
This is very interesting, but it's a bit too much for me right now 🙏

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+      console.error("Failed to save changes in user info", err);
+    })
+    .finally(() => {
+      profileEditPopup.close();
```
` profileEditPopup.close();`

You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server)

**Please, correct it everywhere in the code**

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+    .editProfileImg(input.imageLink)
+    .then((data) => {
+      if (data && data.avatar) {
+        profileImageEl.src = data.avatar;
```
* `UserInfo` should set the entire user data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+      console.error("Failed to update profile image:", err);
+    })
+    .finally(() => {
+      profileEditImagePopup.close();
```
```js

      profileEditImagePopup.close();
      editProfileImgForm.reset();
      formValidators["edit-profile-image-form"].disableButton();
```

These 3 lines of code should be in `then` in line 119

--------------------

You need to **reset** the form only   in a `then` block in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.

gennady-bars replied:
You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.

gennady-bars replied:
**Please, correct it everywhere in the code**

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,236 @@
+    });
+}
+
+function likeClick(cardId, card) {
+  if (!card.isLiked) {
+    api
+      .likeCard(cardId)
+      .then((res) => {
+        card.handleLikeIcon();
+        card.isLiked = res.isLiked;
+      })
+      .catch((err) => {
+        console.error("Failed to like card", err);
```
Nice likes handling

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Mark!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Nice likes handling

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* `UserInfo` should set the entire user data (including the **avatar**)
* You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server)
* You need to **reset** the form only   in a `then` block in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.
* You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

MarkALevy committed:
fix: comments from submit #1

MarkALevy commented:
Hey Gennady!
Thank you for your review. 
I've addressed the comments and pushed the "Needs correcting" changes. Please take a look when you get a chance. Thanks!

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,93 +1,242 @@
+    .editProfileImg(input.imageLink)
+    .then((data) => {
+      if (data && data.avatar) {
+        userInfo.setUserInfo({ userImg: data.avatar });
```
* When I edit the `avatar` , the profile data disappears from the page https://snipboard.io/Nx3OhM.jpg 

Please, add `userName, userJob, userImg` here because `setUserInfo` requieres 3 of them

MarkALevy replied:
fixed

gennady-bars reviewed (commented):
Everything is almost fine, but:

* When I edit the `avatar` , the profile data disappears from the page https://snipboard.io/Nx3OhM.jpg

MarkALevy committed:
fix: profile data disappearing when changing avatar, add _request method to Api.js

MarkALevy commented:
> Everything is almost fine, but:
> 
> * When I edit the `avatar` , the profile data disappears from the page https://snipboard.io/Nx3OhM.jpg

Thank you for your quick review.
I fixed this issue and created the `_request()` method in `Api.js` as you suggested.
Please take a look when you get a chance. Thanks!

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,91 @@
+  };
+
+  likeCard = (cardId) => {
+    return this._requesth(`${this._baseUrl}/cards/${cardId}/likes`, {
```
`_requesth`

* I can't like cards. See the errors https://snipboard.io/cDyt7S.jpg 

It must be `_request`

MarkALevy replied:
oh no! my bad.... stupid typos :)

MarkALevy replied:
fixed

gennady-bars reviewed (commented):
One little thing is left:

* I can't like cards. See the errors https://snipboard.io/cDyt7S.jpg

MarkALevy committed:
fix: typo in Api.likeCard

gennady-bars commented on src/components/Api.js:
```diff
@@ -24,8 +24,9 @@ export default class Api {
+      headers: this._headers,
+    }).then((data) => {
       return data;
     });
```
```js

    }).then((data) => {
      return data;
    });
```

Actually, this has no effect because you get the `data` and return it without changing


```js

  getUserInfo() {
    return this._request(`${this._baseUrl}/users/me`, {
      headers: this._headers,
    });
  }

  getInitialCards() {
    return this._request(`${this._baseUrl}/cards`, {
      headers: this._headers,
    });
  }

```

gennady-bars commented on src/components/Api.js:
```diff
@@ -24,8 +24,9 @@ export default class Api {
-    return this._request(`${this._baseUrl}/cards`, { headers: this._headers })
-    .then((data) => {
+    return this._request(`${this._baseUrl}/cards`, {
+      headers: this._headers,
```
Not to repeat `headers: this._headers` and `baseUrl` in every request, you can put them initially in `_request`

```js

  _request(endpoint, options = {}) {
    const finalOptions = {
      // add the headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };

    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._handleRes);
  }

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    return this._request("/cards");
  }
```

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!

MarkALevy committed:
final touches
